### DISCUSSION TITLE
v8.0.0 (2023-06-04)

### DISCUSSION THREAD

**Discussion by yajo** [copier-org/copier#1174] [General]
### BREAKING CHANGE

- Updates will overwrite existing files always. If you need to select only some files, just use `git mergetool` or `git difftool` after updating.
- Flag `--overwrite/-w` disappeared from `copier update`. It is now implicit.
- To update via API, `overwrite=True` is now required.
- The default update conflict mode is now `inline` instead of `rej`.
- By default, updates now consider 3 lines of context instead of just 1.
- All CLI calls to Copier must now include the subcommand as the 1st argument. For example, `copier` must become now `copier update`; also `copier ./tpl ./dst` must become `copier copy ./tpl ./dst`.
- All flags must go after the subcommand now. For example, `copier -r HEAD update ./dst` must now become `copier update -r HEAD ./dst` or `copier update ./dst -r HEAD`.
- Automatic mode removed. Since now subcommands are required, the automatic mode is removed.
- Deprecated `copier.copy` function is removed. Use `copier.run_copy`, `copier.run_update` or `copier.run_recopy` explicitly as needed.
- default values must be of the same type than the question.

### Feat

- disable unsafe features by default and add `--UNSAFE` switch (#1171)
- basic nixpkgs overlay
- add `recopy` command and function
- support conditional choices (#1010)
- validate default values (#1075)

### Fix

- explain better why an answer casting fails
- **cli**: display subcommand args meaning
- preserver recursive symlinks
- work around Pydantic bug when parsing choices
- skip validating question and generating its default value when its skip condition is met

### Refactor

- overwrite always on updates
- **update**: default to inline markers and 3 lines of context
- remove unused local overrides to answers

<hr /><em>This discussion was created from the release <a href='https://github.com/copier-org/copier/releases/tag/v8.0.0'>v8.0.0 (2023-06-04)</a>.</em>

**Comment by drewoldag** [comment]
Have `tasks` been deprecated now? i.e. commands under `_tasks:` in the copier.yml file? 

Look over the code, it seems like for a `copy` command, if there are tasks associated, then an exception will be raised, copier will stop.

**Reply by drewoldag** [reply]
Ah, ok, `--UNSAFE` is now required if we want to use tasks.

**Reply by drewoldag** [reply]
This is kind of a big deal for our users, I don't think they are going to be excited about having to pass `--UNSAFE`. The task list for our template is simple, `git init --initial-branch=main`, but we also make use of versioning in setuptools, and without a git initialized local repo, it falls over. 

The change here is going to cause a huge headache, and a lot of concern for our users. Please advise what the best path forward here is.

**Reply by sisp** [reply]
See https://github.com/copier-org/copier/issues/1137#issuecomment-1576135992 (and following comments there) for some discussion about this new flag. A registry of trusted templates might be an option. Would this help your users?

**Reply by drewoldag** [reply]
It sounds like a better option than the scary sounding "UNSAFE", but it also sounds like that might be a while before it's available?

**Reply by sisp** [reply]
Not necessarily. Anybody can contribute. :wink: Otherwise, I might be able to take a look next week. But before, we should discuss the implementation in a new issue, e.g. the location and format of the trusted list of templates, URL canonicalization, whether/how we might support wildcards in the URL, and possibly more.

**Reply by carlsylvia** [reply]
I second the concern about requiring an "--UNSAFE" flag to use tasks and extensions.  We would not be able to release a template that required such a tag.  This may render Copier unusable for our company.  You should really think about corporate users when you make these types of changes.  I would recommend you roll that change back.

**Reply by pawamoy** [reply]
@carlsylvia I'm sure Copier's maintainers think about all their users, corporate users comprised, whether they help to ensure the project's sustainability or not. As said in another thread, if v8 is problematic to you you can stay on v7 while a "trusted template" feature is implemented.

**Reply by yajo** [reply]
We're just making the already existing dangers obvious to the users. Before v8, a template could execute any random code in the user machine. Running `copier copy ...` had the same level of risk as `curl ... | bash`. Thus, it was extremely unsafe. Something had to be done.

Is a company eager to use unsafe tools as long as that unsafety is hidden under the carpet? Maybe the problem is the safety policy in the company, and not copier.

Don't get me wrong, I'm not saying this can't be improved. But I appreciate it when tools warn me before potentially breaking my system. Any UX improvements are welcome, but the feature will stay, one way or another.

**Comment by carlsylvia** [comment]
The addition of this flag is very problematic and I recommend you consider rolling this back. My guess is the majority of your users are building templates for in-house use, and your attempt to implement some form of "safeness" is not appropriate. How are you determining/defining what is "safe"?

**Reply by yajo** [reply]
A template that includes tasks or extra extensions is unsafe.

Also one update that will execute migrations.

**Comment by Kludex** [comment]
I didn't read what is written here, but I can't use `copier . .` anymore. Is this going to be fixed?

I don't know the background of this, but it would be cool to have a good error message. 🤷‍♂️

**Reply by yajo** [reply]
The shortcut syntaxes are removed to open room for more subcommands.

**Comment by gerektoolhy** [comment]
Update:


It appears that subcommands are now mandatory.  See my next comment for explanation. 

======


Hey,  I am confused regarding `Flag --overwrite/-w disappeared from copier update. It is now implicit.` change.   It seems that what worked in v7 does not work in v8. See command below: 

```
copier init_project_dir /Users/destination_folder -w -d param_a=value_a -d param_b=value_b 
--
Error: Unknown switch -w
------
copier 8.0.0
```

Could this be an unwanted side effect?    My initial thought was that by default copier executes `update` if  no sub-command is given, however, checking the documentation it says otherwise:
![image](https://github.com/copier-org/copier/assets/131102/de1e0341-0a4d-4206-aaed-500c94ef1f2c)
(see [Generating project](https://copier.readthedocs.io/en/latest/generating/) documentation)

And to update a project one has to explicitly provide subcommand `copier update` (see [Updating a project](https://copier.readthedocs.io/en/latest/updating/) documentation)

**Reply by gerektoolhy** [reply]
Update:

ok, i missed this part:
```
All CLI calls to Copier must now include the subcommand as the 1st argument. For example, copier must become now copier update; also copier ./tpl ./dst must become copier copy ./tpl ./dst.
```

So looks like documentation has to be updated to reflect this, as it is still saying otherwise.

Also, this looks like a very important breaking change, and IMO should be somehow marked more explititly, e.g. perhaps this can be moved to the top of breaking changes.  But maybe thats just me.

**Reply by sisp** [reply]
Thanks for reporting outdated docs! :pray: Let me fix them quickly.

**Reply by gerektoolhy** [reply]
Thank you!   

I also missed explanation on the subcommands, i.e. does `copier source dest` automatically default to `copier copy source dest`?  That would be good to clarify as well!

**Reply by sisp** [reply]
Prior to Copier v8, `copier source dest` was implicitly equivalent to `copier copy source dest`, but we decided to require explicit subcommands like `copy` and `update` and move the flags and switches from the global level to the subcommands (see #1081, #1143).

**Reply by sisp** [reply]
I've submitted a PR that updates the docs: #1192
